File: clients/web/src/app/shared/utils/journal-entry.util.ts
import { formatGridCellValue } from './field-display.util';
import { resolveRecordDisplayLabel } from './lookup-display.util';

export const JE_LINE_PARENT_FIELD = 'journal_entry_id';
export const JE_LINE_ENTITY_CODE = 'JOURNAL_ENTRY_LINE';
export const JE_CREATE_PREFILL_PARAM = 'journal_entry_id';

/** Filter JOURNAL_ENTRY_LINE rows for a parent journal entry. */
export function filterJournalLines(
  lines: Record<string, unknown>[],
  journalEntryId: string,
): Record<string, unknown>[] {
  return lines.filter((row) => String(row[JE_LINE_PARENT_FIELD] ?? '') === journalEntryId);
}

export function journalLineAmount(line: Record<string, unknown>, field: 'debit' | 'credit'): number {
  const value = Number(line[field]);
  return Number.isFinite(value) ? value : 0;
}

export function sumJournalLineAmounts(
  lines: Record<string, unknown>[],
  field: 'debit' | 'credit',
): number {
  return lines.reduce((sum, line) => sum + journalLineAmount(line, field), 0);
}

export function formatJournalLineAmount(
  line: Record<string, unknown>,
  field: 'debit' | 'credit',
  currencyCode = 'USD',
): string {
  const amount = journalLineAmount(line, field);
  if (amount === 0) {
    return '—';
  }
  return formatGridCellValue(field, amount, {
    fieldType: 'currency',
    currencyCode,
  });
}

export function formatJournalLinesTotal(total: number, currencyCode = 'USD'): string {
  if (total === 0) {
    return '—';
  }
  return formatGridCellValue('amount', total, {
    fieldType: 'currency',
    currencyCode,
  });
}

export function resolveAccountLabel(
  accountId: string,
  labels: Record<string, string>,
): string {
  if (!accountId) {
    return '—';
  }
  return labels[accountId] ?? accountId;
}

export function buildAccountLabelMap(
  accounts: Record<string, unknown>[],
): Record<string, string> {
  const map: Record<string, string> = {};
  for (const account of accounts) {
    const id = String(account['id'] ?? '');
    if (id) {
      map[id] = resolveRecordDisplayLabel(account);
    }
  }
  return map;
}

export function journalLineAccountLabel(
  line: Record<string, unknown>,
  labels: Record<string, string>,
): string {
  return resolveAccountLabel(String(line['account_id'] ?? ''), labels);
}

export function canPostJournalEntry(
  entityCode: string,
  record: Record<string, unknown>,
  options: { recordId?: string | null; creatingNew?: boolean } = {},
): boolean {
  const { recordId, creatingNew = false } = options;
  const id = recordId ?? String(record['id'] ?? '');
  if (entityCode !== 'JOURNAL_ENTRY' || !id || creatingNew) {
    return false;
  }
  return String(record['status'] ?? '') === 'draft';
}

export function canVoidJournalEntry(
  entityCode: string,
  record: Record<string, unknown>,
  options: { recordId?: string | null; creatingNew?: boolean } = {},
): boolean {
  const { recordId, creatingNew = false } = options;
  const id = recordId ?? String(record['id'] ?? '');
  if (entityCode !== 'JOURNAL_ENTRY' || !id || creatingNew) {
    return false;
  }
  return String(record['status'] ?? '') === 'posted';
}

export function canAddJournalLine(
  entityCode: string,
  record: Record<string, unknown>,
  options: { recordId?: string | null; creatingNew?: boolean } = {},
): boolean {
  return canPostJournalEntry(entityCode, record, options);
}

File: clients/mobile/lib/utils/journal_entry_util.dart
/// JOURNAL_ENTRY post/void + line helpers — web `journal-entry.util.ts` parity.

import 'field_display.dart';
import 'lookup_display.dart';

const journalLineParentField = 'journal_entry_id';
const journalLineEntityCode = 'JOURNAL_ENTRY_LINE';
const journalEntryCreatePrefillParam = 'journal_entry_id';

List<Map<String, dynamic>> filterJournalLines(
  List<Map<String, dynamic>> lines,
  String journalEntryId,
) {
  return lines
      .where((row) => '${row[journalLineParentField] ?? ''}' == journalEntryId)
      .toList();
}

double journalLineAmount(Map<String, dynamic> line, String field) {
  final value = double.tryParse('${line[field] ?? ''}');
  return value ?? 0;
}

double sumJournalLineAmounts(
  List<Map<String, dynamic>> lines,
  String field,
) {
  return lines.fold(0.0, (sum, line) => sum + journalLineAmount(line, field));
}

String formatJournalLineAmount(
  Map<String, dynamic> line,
  String field, {
  String? locale,
  String currencyCode = 'USD',
}) {
  final amount = journalLineAmount(line, field);
  if (amount == 0) {
    return '—';
  }
  return formatGridCellValue(
    field,
    amount,
    locale: locale,
    fieldType: 'currency',
    currencyCode: currencyCode,
  );
}

String formatJournalLinesTotal(
  double total, {
  String? locale,
  String currencyCode = 'USD',
}) {
  if (total == 0) {
    return '—';
  }
  return formatGridCellValue(
    'amount',
    total,
    locale: locale,
    fieldType: 'currency',
    currencyCode: currencyCode,
  );
}

String resolveAccountLabel(String accountId, Map<String, String> labels) {
  if (accountId.isEmpty) {
    return '—';
  }
  return labels[accountId] ?? accountId;
}

Map<String, String> buildAccountLabelMap(List<Map<String, dynamic>> accounts) {
  final map = <String, String>{};
  for (final account in accounts) {
    final id = '${account['id'] ?? ''}';
    if (id.isNotEmpty) {
      map[id] = resolveRecordDisplayLabel(account);
    }
  }
  return map;
}

String journalLineAccountLabel(
  Map<String, dynamic> line,
  Map<String, String> labels,
) {
  return resolveAccountLabel('${line['account_id'] ?? ''}', labels);
}

bool canPostJournalEntry(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
}) {
  final id = recordId ?? '${record['id'] ?? ''}';
  if (entityCode != 'JOURNAL_ENTRY' || id.isEmpty || creatingNew) {
    return false;
  }
  return '${record['status'] ?? ''}' == 'draft';
}

bool canVoidJournalEntry(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
}) {
  final id = recordId ?? '${record['id'] ?? ''}';
  if (entityCode != 'JOURNAL_ENTRY' || id.isEmpty || creatingNew) {
    return false;
  }
  return '${record['status'] ?? ''}' == 'posted';
}

bool canAddJournalLine(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
}) {
  return canPostJournalEntry(
    entityCode,
    record,
    recordId: recordId,
    creatingNew: creatingNew,
  );
}
